### PR TITLE
feat(notion-react): JSX-driven page ops + sub-page create/update/archive/move

### LINKED ISSUE #618: Epic: JSX-driven page ops (root metadata + sub-pages)
# Epic — JSX-driven page ops in `@overeng/notion-react`

Replaces the v0 "two-step pattern" note with a principled, long-term design that treats pages and sub-pages as first-class surfaces. Tracks all work needed to close the gap: root page metadata (`<Page title icon cover>`), `<ChildPage>` create / update / archive / move with their own children, and the latent bug where `<ChildPage title>` changes currently emit a rejected `blocks.update` call.

## Progress (`schickling/2026-04-23-notion-react-followup2`)

- [x] **VRS updates** — vision, requirements (A06–A10, T06–T09, R24–R30, R21a), spec (page-ops layer, DQ6/DQ7) — `1474a129`
- [x] **Phase 1** — `PageOp` types, `SyncResult.pages`, fallback reasons, additive `nodeKind` on `CacheNode` — `2a355594`
- [x] **Phase 2** — latent-bug fix: `<ChildPage title>` change → `pages.update` (both emitter sites) — `3a362362`
- [x] **Phase 3a** — widen `PageProps` / `ChildPageProps`; stateful `FakePage` mock; `'page_root'` host type — `009744bb`
- [x] **Phase 3b** — diff emits `createPage`/`updatePage`/`archivePage`/`movePage`; driver executes them; partial-failure archive; icon/cover normalization; 9 new unit tests + e2e skeleton — `5d0374da`
- [x] **Phase 3c** — recursive sub-page reconciliation; cache v3 bump; per-page `blockKey` namespace — `9273d705`
- [ ] **Phase 3d** — live e2e hardening: full matrix against real Notion, property-based idempotency
- [ ] **Phase 4** — polish: `ensureSiblingOrder` re-fetch, partial-failure edge cases
- [ ] **Phase 5** — remaining test matrix (regression + property-based)
- [x] **Phase 6** — stories, docs (cookbook, concepts, migration, api.md), CHANGELOG finalize — `ab54b80e`
- [ ] **Merge / PR** — final CI green, PR description, close the epic

## Post-3d follow-up (dispatched from options research)

Research docs: [`tmp/notion-618/options-ordering.md`](../tree/schickling/2026-04-23-notion-react-followup2/tmp/notion-618/options-ordering.md), [`options-database-pages.md`](../tree/schickling/2026-04-23-notion-react-followup2/tmp/notion-618/options-database-pages.md), [`options-small-gaps.md`](../tree/schickling/2026-04-23-notion-react-followup2/tmp/notion-618/options-small-gaps.md).

- [x] **Phase 4a** — T08 sequential-by-default (`793adcb7`): flip default for same-parent `<ChildPage>` sibling creates; drop the planned `ensureSiblingOrder` re-fetch (structural determinism over latency).
- [x] **Phase 4b** — Cover / icon `null` sentinel (`5b6bc471`): `cover={null}` / `icon={null}` emits `updatePage({cover: null})` / `({icon: null})`. Matches Notion API shape.
- [x] **Phase 4c** — Web renderer icon/cover (`ab63e9f3`): render `renderPageIcon` / `renderPageCover` in `src/web/blocks.tsx`; stories cover emoji / external / file / absent.
- [x] **Phase 4d** — DQ7 opt-in `reorderSiblings` (`fd65796c`): move-roundtrip primitive (2N moves, ids preserved) for same-parent reorder. Holding-parent auto-provisioned or caller-supplied.
- [ ] PR review follow-through: PR #623 Codex P1 fixes landed (`97585d89`) — movePage descendants + resolveInlineChildrenIds alignment.

### Deferred to separate epic

- **DQ6** `<DbRow>` — database-parented pages with explicit schema. Design in `options-database-pages.md`. Scope is an L-sized epic in its own right; fold into a follow-up issue once a concrete caller asks.

## Why

- Today `<ChildPage>` is reference-only; creating a sub-page requires callers to drop to the imperative `NotionPages.create` API and pass the id back as `blockKey`. Composition is broken across the page boundary.
- Root `<Page>` carries no metadata — title / icon / cover must be set by hand outside JSX, even though the sync driver already has a `pageId` in hand.
- `<ChildPage title>` changes currently emit `NotionBlocks.update({child_page:{title}})`, which the Notion API rejects with `validation_error`. Confirmed via live API probe (see `tmp/notion-618/experiments/`). **Fixed in phase 2.**

## Architecture — Option B (per-page sync boundary)

- Each rendered page (root + every nested `<ChildPage>`) is its own sync boundary with its own `blockKey` namespace, its own cache subtree, and its own `OpBuffer`.
- `DiffOp` widens to `BlockOp | PageOp`. `PageOp = createPage | updatePage | archivePage | movePage`.
- `<ChildPage>` title / icon / cover changes route exclusively through `NotionPages.update` — never the block API.
- `<ChildPage>` removal emits `archivePage` (`pages.update {in_trash:true}`), not `blocks.delete`.
- Reparenting a retained `<ChildPage>` uses `NotionPages.move` (id-preserving); intra-parent reorder is unsupported in v1 (DQ7).
- Cache schema `CACHE_SCHEMA_VERSION` bumps 2→3 with a new `nodeKind: 'block' | 'page'` discriminator; migration goes through the existing `"schema-mismatch"` fallback.
- New fallback reasons: `"page-missing"`, `"page-archived"`, `"partial-page-create"`.

VRS on this branch reflects the target architecture:
- `packages/@overeng/notion-react/docs/vrs/vision.md` — S6 added; "Not a general-purpose block database ORM" widened.
- `packages/@overeng/notion-react/docs/vrs/requirements.md` — A06–A10, T06–T09, R24–R30, R21a added.
- `packages/@overeng/notion-react/docs/vrs/spec.md` — Page-ops layer section, cache v3 schema, fallback table rows, DQ6/DQ7 added.

## Empirical findings (live API probes)

Full transcript in `tmp/notion-618/experiments/` on branch `schickling/2026-04-23-notion-react-followup2`.

- `child_page` is not appendable via `blocks.children.append` — only `pages.create` materializes a sub-page; the auto-created `child_page` block has the same id as the new page.
- `blocks.update` rejects any `child_page` body — title / icon / cover must go through `pages.update`.
- Inline `children` on `pages.create` allowed up to depth 2; deeper requires follow-up `blocks.children.append`. 100-child cap per request applies to both.
- `pages.move` with `{parent:{page_id}}` is real, id-preserving, non-destructive.
- Archived pages: `pages.retrieve` / `blocks.retrieve` still 200; only `blocks.children.list` returns 404. Un-archive restores all children.
- Icon response-shape ≠ request-shape: built-in SVG URLs resolve to `{type:"icon",icon:{name,color}}`; unknown `custom_emoji` ids silently drop to `null`. Cache must hash the response-normalized shape.
- Cover accepts only `external` / `file_upload`. Emoji/custom_emoji covers → 400.
- Title length cap is 2000 chars **per rich_text span** (multi-span escape hatch).
- Concurrent sibling creates: no ordering guarantee. Re-fetch children when order matters.
- No idempotency / request-token primitive. Only `x-notion-request-id` for debugging.
- Strong consistency: `pages.create` with inline children → immediate `blocks.children.list` is fully populated.

## Phased implementation plan

Each phase lands green CI (`dt check:all`) with tests. Phases are independently mergeable.

### Phase 0 — Latent-bug fix-first
Superseded: folded into Phase 2.

### Phase 1 — Schema + types ✅ `2a355594`
Additive. No behavior change.

### Phase 2 — Latent-bug fix ✅ `3a362362`
`<ChildPage title>` change → `pages.update`, both emitter sites, minimal mock stub, unit + e2e regression.

### Phase 3a — Props + host-config + FakePage ✅ `009744bb`
- `PageProps` / `ChildPageProps` carry `title` (string | span[]), `icon`, `cover`; `ChildPage` gains `children`.
- `<Page>` renders as `'page_root'` host type (not Fragment).
- `<ChildPage>` marked `nodeKind: 'page'` in host-config.
- `issueBlockUpdate` widened to route icon/cover + rich-text titles through `pages.update`.
- Stateful `FakePage` mock: POST/PATCH/GET `/v1/pages/{id}`, POST `/v1/pages/{id}/move`, archived-page 404 on children list.

### Phase 3b — Diff emits + driver executes PageOps ✅ `5d0374da`
- `CandidateNode.nodeKind`; LCS predicate `(key, nodeKind, type)`.
- Emits `createPage` (with inline packing up to depth 2 / 100 blocks), `updatePage`, `archivePage`, `movePage`.
- `indexCachePages` for move detection across the sibling set (identity preserved via `blockKey`).
- Apply order: root `updatePage` → root block ops → `createPage` + tail blocks → retained `updatePage` → `movePage` → `archivePage`.
- Partial-failure: create succeeds + tail fails → archive the new page + `fallbackReason: 'partial-page-create'`.
- `icons.ts`: `projectIcon` / `normalizeIcon` fold built-in-SVG rewrite into canonical hash form.
- `SyncEvent.PageOpIssued` / `PageOpApplied`.
- 9 new unit tests in `sync-page-ops.unit.test.tsx`; initial `sub-pages.e2e.test.tsx` scaffold.
- `ChildPageProps` composes `BlockKey` so sibling reshuffles are distinguishable from parallel renames.

### Phase 3c — Recursive sub-page reconciliation (in progress)
- `buildCandidateTree` walks into `<ChildPage>` children.
- Sync driver recurses per-page; each sub-page has its own `blockKey` namespace + cache subtree.
- `CACHE_SCHEMA_VERSION` bumps 2→3. v2 caches fall through `"schema-mismatch"`.
- Deep-tree unit tests: idempotent resync, nested-page mutation, sub-page in sub-page, blockKey isolation, depth-3 inline splitting, >100 children under new sub-page.

### Phase 3d — Live e2e hardening
- Run `sub-pages.e2e.test.tsx` against real Notion (demo workspace token in 1P at `Private/Notion/Effect Utils Notion Workspace/u7q2coiqw5wdt4ab33yia3g4w4`).
- Expand matrix: root metadata multi-field change, archive cascade, rapid resync idempotence, concurrent sibling create ordering, >100 children, depth-3 inline split, cache v2→v3 migration, move across parents.
- Property-based: random JSX trees with `<ChildPage>` → two syncs produce same server state and zero ops on second run.

### Phase 4 — Polish
- `ensureSiblingOrder` opt-in re-fetch after concurrent creates.
- Any partial-failure edge case exposed by e2e.

### Phase 5 — Remaining test matrix
- Regression suite extensions (`sync-perf.regression`).
- Mock-client round-trip property tests.

### Phase 6 — Stories, docs, demo, changelog
- `src/web/pages.stories.tsx`: root metadata + nested `<ChildPage>` story.
- `README.md` sub-pages section (JSX-driven).
- `docs/api.md`, `docs/cookbook/sub-page-creation.md`, `docs/concepts/page-boundaries.md`, `docs/migration.md` (v2→v3).
- Root `CHANGELOG.md` — finalize `[Unreleased]` (entries added incrementally by phases).

### Phase 7 — Bonus (time permitting)
- Database-parented `<ChildPage>` (DQ6).
- Rate-limit body-based back-pressure.

## Acceptance criteria (maps to R24–R30)

- [x] Latent bug: `<ChildPage title>` change → `pages.update` regression covered (phase 2).
- [x] R24: Root `<Page>` metadata change → exactly one `pages.update` (phase 3b).
- [x] R25: Sub-page create / update / archive via JSX — no block-API calls touch page metadata (phase 3b).
- [x] R26: Per-page sync boundary (independent blockKey namespaces, independent cache subtrees) — phase 3c.
- [x] R27: `<ChildPage>` reparent uses `pages.move`; id preserved (phase 3b).
- [x] R28: Partial failure archives orphan; next sync reconciles by id (phase 3b).
- [x] R29: `SyncResult.pages` + `SyncEvent` variants (phases 1 + 3b).
- [x] R30: Cache v2→v3 via `"schema-mismatch"`; new fallback reasons wired — phases 1 + 3b + 3c.
- [ ] `dt check:all` green on the branch.
- [ ] CHANGELOG `[Unreleased]` entry finalized (entries added per-phase).

## Risks & mitigations

- **Recursive sync complexity** — concentrate complexity in `sync.ts`; ship Phase 3c as one atomic commit; keep checkpointing per-page so failures are recoverable.
- **Cache migration** — v2→v3 is transparent via `"schema-mismatch"` fallback; no hard errors.
- **Icon response-shape drift** — response-normalization before hashing; component-boundary validation for `custom_emoji`. ✅ in phase 3b.
- **Large refactor in `host-config.ts`** — guard behind `nodeKind` branching; don't remove block paths. ✅ in phase 3a.

## Out of scope

- Database-parented pages (deferred, DQ6).
- `is_locked`, `erase_content` (rare, caller can still hit `pages.update` directly).
- Intra-parent sibling-page reorder (DQ7 — Notion API doesn't support it cheaply).
- Async uploads inside newly created sub-pages (still v0.2 per R15).

## Refs

- PR #584 (incubation)
- PR #621 (post-#584 cleanup)
- Branch: `schickling/2026-04-23-notion-react-followup2`
- Live probes: `tmp/notion-618/experiments/`
- Proposal + code-map: `tmp/notion-618/proposal.md`, `tmp/notion-618/code-map.md`
- Assets audit: `tmp/notion-618/assets-audit.md`

## Phase 5 — Storybook hardening (landed)

Audit + research docs: [`tmp/notion-618/storybook-gaps.md`](../tree/schickling/2026-04-23-notion-react-followup2/tmp/notion-618/storybook-gaps.md), [`storybook-5c-options.md`](../tree/schickling/2026-04-23-notion-react-followup2/tmp/notion-618/storybook-5c-options.md), [`storybook-d4-visual-sweep.md`](../tree/schickling/2026-04-23-notion-react-followup2/tmp/notion-618/storybook-d4-visual-sweep.md).

- [x] **5a** — `<ChildPage>` clickable `<a>` + `children` passthrough; stories for `Pdf`/`Audio`/`Video`/`File` + passthrough blocks; `CanonicalPage` story; emoji font-stack + inline icon sizing CSS fixes (`dcf3972a`).
- [x] **5b-1** — `NotionUrlProvider` context (sync, optional, silent-miss per D1) wired through `<ChildPage>` / `<LinkToPage>` / page-`<Mention>`; `.storybook/preview.tsx` cross-story resolver (`aca3cba8`).
- [x] **5b-2** — `defaultOpen` on `<Toggle>` + `body?` on toggleable `<Heading>` (per D2). Web-mirror-only; never projected to Notion (`82c483d2`).
- [x] **5b-3** — `argTypes` + `tags: ['autodocs']` on interactive stories; empty/error-state stories (per D3) (`71f36d9b`).

Open questions resolved via empirical research + live visual sweep on Netlify preview:

- D1: URL resolver = sync `(ref: {pageId, blockId?}) => string | undefined`; optional context; silent fallback — mirrors `UploadRegistry` precedent.
- D2: Toggle stays Notion-faithful (default closed); new `defaultOpen` prop opts stories into expansion.
- D3: Skip MDX; adopt `argTypes` + `autodocs` per repo convention.
- D4: Visual sweep surfaced two live-only gaps (emoji tofu + inline icon 0×0 collapse) — both fixed in 5a.

<details>
<summary>Posted on behalf of @schickling</summary>

- `agent_name`: ⚙️ cl1-zinc
- `agent_session_id`: 06d6166e-8c2e-4d5e-aae9-9c575c0f414c
- `agent_tool`: Claude Code
- `agent_tool_version`: 2.1.114 (Claude Code)
- `agent_runtime`: Claude Code 2.1.114 (Claude Code)
- `agent_model`: unknown
- `worktree`: effect-utils/schickling/2026-04-23-notion-react-followup2
- `machine`: dev3
</details>

### PR DESCRIPTION
**Pull request by schickling-assistant**
## Summary

Implements JSX-driven page ops for `@overeng/notion-react` — root `<Page>` metadata (title/icon/cover) and `<ChildPage>` create/update/archive/move with their own children. Fixes the latent bug where `<ChildPage title>` changes emitted a rejected `blocks.update` call. Closes #618.

See issue #618 for the full phased plan, empirical findings (`tmp/notion-618/experiments/`), and acceptance criteria. VRS updated in lockstep (vision S6, requirements A06–A10 / T06–T09 / R24–R30 / R21a, spec page-ops section).

## Status

**Draft** — still landing:

- [x] Phase 1: type plumbing (`PageOp`, `SyncResult.pages`, additive `nodeKind`)
- [x] Phase 2: latent-bug fix — `<ChildPage title>` change → `pages.update`
- [x] Phase 3a: props widened; `'page_root'` host type; stateful `FakePage` mock
- [x] Phase 3b: diff emits + driver executes `createPage`/`updatePage`/`archivePage`/`movePage`; partial-failure archive; icon/cover normalization
- [x] Phase 3c: recursive sub-page reconciliation + cache v3 bump
- [ ] Phase 3d: live e2e hardening (full matrix against real Notion + property-based)
- [ ] Phase 4: polish (`ensureSiblingOrder`, etc.)
- [x] Phase 6: stories, docs (cookbook, concepts, migration, api.md), CHANGELOG finalize

## Rationale

The block/page API boundary is real and not glossable: `child_page` blocks are not creatable via `blocks.children.append`, and `blocks.update` rejects any `child_page` body. Empirically verified (see `tmp/notion-618/experiments/`). Option B — per-page sync boundary — concentrates complexity in one place (`sync.ts` recursion) rather than smearing it across op-buffer, diff, and cache. Alternatives (Option A: heterogeneous ops in one pipeline; Option C: stacked reconcilers) were considered and rejected on clarity grounds (see `tmp/notion-618/proposal.md`).

Tradeoffs worth flagging:

- **No cross-page op batching** (T07): page ops are always individual requests. Acceptable because page ops are rare and per-page scope is simpler.
- **Partial-failure archive** (T06): a sub-page created mid-flight whose block tail fails is archived. No transactional alternative exists (Notion has no idempotency primitive).
- **Intra-parent sibling-page reorder is unsupported** (DQ7): Notion's `pages.move` only accepts a new parent. Cross-parent moves work; same-parent reorder would require archive+recreate, which loses the id. Documented, not silent.
- **Database-parented pages deferred** (DQ6): `<ChildPage>` currently targets page parents only. Prop shape is designed to allow `properties` + discriminated-union `parent` without a schema change.

## Test plan

- [x] Unit: `sync-page-ops.unit.test.tsx` (9 scenarios: create, idempotent resync, root metadata, sub-page metadata, archive, sibling-reshuffle, nodeKind-LCS boundary, partial-failure, cover-shape)
- [x] Unit: `sync-mock.unit.test.tsx` (title/icon/cover routing through `pages.update`)
- [x] Unit: `mock-client.unit.test.ts` (FakePage CRUD contract)
- [x] E2E scaffold: `sub-pages.e2e.test.tsx` — create/rename/icon/idempotence/archive (one scenario)
- [ ] E2E expansion (Phase 3d): cover change, reparent, deep tree, >100 children, depth-3 inline split, concurrent sibling create, cache v2→v3 migration, multi-field root metadata
- [ ] Property-based: random JSX trees with `<ChildPage>` → two syncs produce same server state and zero ops on second run
- [ ] Full CI green (`dt check:all`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Bugs surfaced by phase 3d — fixed

1. ✅ **Mixed-root idempotency regression (phase 3c)** (fixed `bf704d2e`) — `<Page>` with mixed `<ChildPage>` + block siblings throws `candidateToCache: unresolved blockId` on the second sync. Repro: `sync-property.unit.test.tsx > sync() childpage-idempotency bug` (currently `it.skip`). Breaks S1 / R04 for a common tree shape.
2. ✅ **Cross-parent `pages.move` race (phase 3b)** (fixed `5f02f177`) — moving a `<ChildPage>` between two sibling parents in one sync leaves server state non-deterministic (archivePage and movePage race on the same id). The e2e `reparent` scenario is `it.skip`-ed pending this fix. Breaks R27.
3. 🔶 **Cover `set → unset` is a no-op** — design question, not fixed this round. — dropping the `cover` prop does not emit a clear. Needs a product-direction decision: honor absence as clear (R24 spirit) or require explicit `cover={null}` sentinel. Not blocking the core flow.

Property-based idempotency suite (`numRuns: 30`) and the cross-parent reparent e2e are now un-skipped and green. The latent nested-scope risk was probed (`98a35b7d`) and confirmed to be a non-issue — nested mixed-siblings reconcile correctly.

## Limitations

All deliberate out-of-scope behaviour and deferrals are documented in
[`packages/@overeng/notion-react/docs/limitations.md`](../blob/schickling/2026-04-23-notion-react-followup2/packages/@overeng/notion-react/docs/limitations.md)
(added `d551e3b2`). Highlights: no database-parented pages (DQ6),
intra-parent sibling-page reorder (DQ7), cover `set → unset` no-op
(open design question), no cross-page op batching (T07),
concurrent-sibling ordering non-authoritative (T08).

## Phase 4 follow-up (landed this round)

- [x] 4a — sequential same-parent `<ChildPage>` creates (`793adcb7`). T08 rewritten: JSX order is authoritative.
- [x] 4b — `cover={null}` / `icon={null}` clear sentinels (`5b6bc471`). Null-vs-absent distinct hashing; `updatePage({cover:null})` emitted on set→null.
- [x] 4c — web mirror renders icon/cover (`ab63e9f3`). New stories for variants.
- [x] 4d — opt-in `reorderSiblings` via `pages.move` roundtrip (`fd65796c`). Per-sync ephemeral holding parent or caller-supplied. 2N moves per burst.

## Phase 5 — Storybook hardening (landed)

Source + live-visual audit, then research-backed decisions. Full synthesis in `tmp/notion-618/storybook-*.md`.

- [x] 5a — ChildPage clickable `<a>` + `children` rendered in DOM; stories for `Pdf`/`Audio`/`Video`/`File` + passthrough blocks; `CanonicalPage` story; emoji font-stack + inline icon sizing (`dcf3972a`).
- [x] 5b-1 — `NotionUrlProvider` context (sync, optional, silent-miss); wired through `<ChildPage>`, `<LinkToPage>`, page-`<Mention>`; `.storybook/preview.tsx` cross-story resolver (`aca3cba8`).
- [x] 5b-2 — `defaultOpen` on `<Toggle>` + `body?` on toggleable `<Heading>` (web-mirror-only; never projected to Notion) (`82c483d2`).
- [x] 5b-3 — `argTypes` + `tags: ['autodocs']` + empty/error-state stories (`71f36d9b`).

Test count: 299 → 314. Full suite green at every commit.

<details>
<summary>Posted on behalf of @schickling</summary>

- `agent_name`: ⚙️ cl1-zinc
- `agent_session_id`: 06d6166e-8c2e-4d5e-aae9-9c575c0f414c
- `agent_tool`: Claude Code
- `agent_tool_version`: 2.1.114 (Claude Code)
- `agent_runtime`: Claude Code 2.1.114 (Claude Code)
- `agent_model`: unknown
- `worktree`: effect-utils/schickling/2026-04-23-notion-react-followup2
- `machine`: dev3
</details>

### CODE REVIEW

File: packages/@overeng/notion-react/src/renderer/sync-diff.ts [hunk 1]
```diff
@@ -281,6 +403,82 @@ const diffChildren = (
...
+        prevRef = prior.blockId
+        continue
+      }
+      // Unretained: either brand-new or a cross-parent move.
+      const moveSource = ctx.pagesByKey.get(cand.key)
+      if (moveSource !== undefined && !ctx.claimedMoves.has(moveSource.blockId)) {
+        ctx.claimedMoves.add(moveSource.blockId)
+        cand.blockId = moveSource.blockId
+        ops.push({
+          kind: 'movePage',
```
[the remark this answers is not in the document]
Comment (schickling-assistant) [reply packages/@overeng/notion-react/src/renderer/sync-diff.ts:559]: Fixed in 97585d89. The movePage branch now recurses via `diffChildren(moveSource.blockId, moveSource.children, cand.children, …)` with `scopePageId = moveSource.blockId`, mirroring the retained-page branch. Regression test: cross-parent move with a nested ChildPage + Paragraph descendant now survives a third sync (previously threw on the unresolved-blockId path).

<details>
<summary>Posted on behalf of @schickling</summary>

- `agent_name`: ⚙️ cl1-zinc
- `agent_session_id`: 06d6166e-8c2e-4d5e-aae9-9c575c0f414c
- `agent_tool`: Claude Code
- `agent_tool_version`: 2.1.114 (Claude Code)
- `agent_runtime`: Claude Code 2.1.114 (Claude Code)
- `agent_model`: unknown
- `worktree`: effect-utils/schickling/2026-04-23-notion-react-followup2
- `machine`: dev3
</details>

File: packages/@overeng/notion-react/src/renderer/sync.ts [hunk 2]
```diff
@@ -646,6 +673,68 @@ const resolveAbsorbedTmpIds = (
...
+          Effect.mapError(
+            (cause) => new NotionSyncError({ reason: 'notion-retrieve-failed', cause }),
+          ),
+        )
+        const liveArr = Chunk.toReadonlyArray(live)
+        for (let i = 0; i < nodes.length && i < liveArr.length; i++) {
+          const cand = nodes[i]!
+          const server = liveArr[i]!
+          if (cand.nodeKind === 'page') continue
+          if (cand.blockId !== undefined && cand.blockId.startsWith('tmp-')) {
```
[the remark this answers is not in the document]
Comment (schickling-assistant) [reply packages/@overeng/notion-react/src/renderer/sync.ts]: Fixed in 97585d89. `resolveInlineChildrenIds` now tracks `liveIdx` independently and advances only on block candidates; page siblings no longer shift alignment. Regression test: page-before-block candidate order round-trips with zero ops.

<details>
<summary>Posted on behalf of @schickling</summary>

- `agent_name`: ⚙️ cl1-zinc
- `agent_session_id`: 06d6166e-8c2e-4d5e-aae9-9c575c0f414c
- `agent_tool`: Claude Code
- `agent_tool_version`: 2.1.114 (Claude Code)
- `agent_runtime`: Claude Code 2.1.114 (Claude Code)
- `agent_model`: unknown
- `worktree`: effect-utils/schickling/2026-04-23-notion-react-followup2
- `machine`: dev3
</details>

**Comment by schickling** [comment]
@codex

File: packages/@overeng/notion-react/src/renderer/sync.ts [hunk 3]
```diff
@@ -1284,74 +1408,615 @@ export const sync = (
...
+    // applyDiff pass so per-parent coalescing stays local; the shared
+    // `working` cache is already indexed over every nested page subtree so
+    // checkpointAppended lands ops in the right place.
+    const createPageTmpIds = new Set(
+      pageOps.flatMap((op) => (op.kind === 'createPage' ? [op.tmpPageId] : [])),
+    )
+    for (const [scope, scopedOps] of pageScopedBlockPlan) {
+      if (createPageTmpIds.has(scope)) continue // tail block ops for a root-scope createPage already applied inside runCreatePage
+      yield* applyDiff(scopedOps, idMap, checkpointAppended, o11y, priorHashById)
+    }
```
[the remark this answers is not in the document]
Comment (schickling-assistant) [reply packages/@overeng/notion-react/src/renderer/sync.ts]: Fixed in 45ed59d7. The retained sub-page scope now walks the plan in candidate order and interleaves scoped block ops with nested `createPage`/`movePage` ops (same pattern as the root scope), so `[<ChildPage>, <Paragraph>]` under a retained sub-page lands as `[child_page, paragraph]` on the server. Regression test asserts server-side order via the fake client after a second sync (previously would silently land in reverse).


<details>
<summary>Posted on behalf of @schickling</summary>

| field | value |
| --- | --- |
| `agent_name` | 🐤 cl1-finch |
| `agent_session_id` | e0c45463-0711-445a-a002-1b4e7cbd6202 |
| `agent_tool` | Claude Code |
| `agent_tool_version` | 2.1.118 (Claude Code) |
| `agent_runtime` | Claude Code 2.1.118 (Claude Code) |
| `agent_model` | claude-opus-4-7 |
| `worktree` | effect-utils/schickling/2026-04-23-notion-react-followup2 |
| `machine` | dev3 |
| `tooling_profile` | dotfiles@f937ca8-dirty |
</details>

File: packages/@overeng/notion-react/src/renderer/sync-diff.ts [hunk 4]
```diff
@@ -375,23 +741,247 @@ const emitAppendsForNew = (parentId: string, children: CandidateNode[], ops: Dif
...
+/**
+ * Index every page-kind entry in the prior cache tree by blockKey. Used by
+ * `diffChildren` to detect cross-parent moves: an unretained candidate page
+ * with a matching key becomes `movePage` instead of archive+create.
+ */
+const indexCachePages = (cache: CacheTree): Map<string, CacheNode> => {
+  const out = new Map<string, CacheNode>()
+  const walk = (nodes: readonly CacheNode[]): void => {
+    for (const n of nodes) {
+      if (n.nodeKind === 'page') out.set(n.key, n)
```
[the remark this answers is not in the document]
Comment (schickling-assistant) [reply packages/@overeng/notion-react/src/renderer/sync-diff.ts]: Fixed in 45ed59d7. `indexCachePages` now detects blockKey collisions and omits them from the map; `pagesByKey.get(cand.key)` returns `undefined` in the ambiguous case and the diff falls back to archive + createPage (safe in all cases) instead of misrouting the move. Regression test: two branches share a blockKey, removing one and adding a new branch with the same key no longer archives/moves the retained page.


<details>
<summary>Posted on behalf of @schickling</summary>

| field | value |
| --- | --- |
| `agent_name` | 🐤 cl1-finch |
| `agent_session_id` | e0c45463-0711-445a-a002-1b4e7cbd6202 |
| `agent_tool` | Claude Code |
| `agent_tool_version` | 2.1.118 (Claude Code) |
| `agent_runtime` | Claude Code 2.1.118 (Claude Code) |
| `agent_model` | claude-opus-4-7 |
| `worktree` | effect-utils/schickling/2026-04-23-notion-react-followup2 |
| `machine` | dev3 |
| `tooling_profile` | dotfiles@f937ca8-dirty |
</details>
